File: backend/src/controllers/orderController.ts
import { type Context } from "hono";
import {
  BaseController,
  controller,
  route,
  httpGet,
  httpQuery,
  httpPost,
  // httpPut,
  // httpDelete,
  fromBody,
  fromRoute,
  fromQuery,
  producesResponseType,
  authorize,
} from "#/framework/mvc";
import type { IOrderService } from "#/core/services/orderService";
import type { IOrderApi } from "@hc-management/shared/api";
import {
  OrderListRequestDto,
  OrderListResponseDto,
  OrderUpsertRequestDto,
  type OrderDetailResponseDto
} from "@hc-management/shared/dtos";

@controller
@route("/api/orders")
@authorize
export class OrderController extends BaseController implements IOrderApi {
  private readonly orderService: IOrderService;

  public constructor(httpContext: Context) {
    super(httpContext);
    this.orderService = this.serviceProvider.getRequiredService("orderService");
  }

  @httpQuery("/")
  @fromQuery(0, OrderListRequestDto)
  @producesResponseType<OrderListResponseDto>(200)
  public async getListAsync(requestDto: OrderListRequestDto): Promise<OrderListResponseDto> {
    return this.ok(await this.orderService.getListAsync(requestDto));
  }

  @httpGet("/:id{[0-9]+}")
  @fromRoute(0, "id", "number")
  @producesResponseType<OrderDetailResponseDto>(200)
  @producesResponseType(401)
  @producesResponseType(403)
  @producesResponseType(404)
  public async getDetailAsync(id: number): Promise<OrderDetailResponseDto> {
    return this.ok(await this.orderService.getDetailAsync(id));
  }

  @httpPost("/")
  @fromBody(0, OrderUpsertRequestDto)
  @producesResponseType<number>(201)
  @producesResponseType(400)
  @producesResponseType(401)
  @producesResponseType(403)
  @producesResponseType(422)
  public async createAsync(requestDto: OrderUpsertRequestDto): Promise<number> {
    const createdId = await this.orderService.createAsync(requestDto);
    return this.created(`/api/menu-category/${createdId}`, createdId);
  }

  // @httpPut("/:id{[0-9]+}")
  // @fromRoute(0, "id", "number")
  // @fromBody(1, OrderUpsertRequestDto)
  // @producesResponseType(200)
  // @producesResponseType(400)
  // @producesResponseType(401)
  // @producesResponseType(403)
  // @producesResponseType(404)
  // @producesResponseType(422)
  // public async updateAsync(id: number, requestDto: OrderUpsertRequestDto): Promise<void> {
  //   await this.orderService.updateAsync(id, requestDto);
  //   return this.ok();
  // }

  // @httpDelete("/:id{[0-9]+}")
  // @fromRoute(0, "id", "number")
  // @producesResponseType(200)
  // @producesResponseType(401)
  // @producesResponseType(403)
  // @producesResponseType(404)
  // @producesResponseType(422)
  // public async deleteAsync(id: number): Promise<void> {
  //   await this.orderService.deleteAsync(id);
  //   return this.ok();
  // }
}

File: backend/Api/Controllers/OrderController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using HCManagement.Core.Features.Orders;

namespace HCManagement.Api.Controllers;

[ApiController]
[Route("api/orders")]
[Authorize]
public class OrderController : ControllerBase
{
    #region Fields
    private readonly IOrderService _service;
    #endregion

    #region Constructors
    public OrderController(IOrderService service)
    {
        _service = service;
    }
    #endregion

    #region Methods
    [HttpPost("list")]
    [ProducesResponseType<OrderListResponseDto>(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    public async Task<IActionResult> GetList([FromBody] OrderListRequestDto requestDto)
    {
        return Ok(await _service.GetListAsync(requestDto));
    }

    [HttpGet("{id:int}")]
    [ProducesResponseType<OrderDetailResponseDto>(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> GetDetail([FromRoute] int id)
    {
        return Ok(await _service.GetDetailAsync(id));
    }

    [HttpPost]
    [ProducesResponseType<OrderDetailResponseDto>(StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
    public async Task<IActionResult> Create([FromBody] OrderUpsertRequestDto requestDto)
    {
        OrderDetailResponseDto responseDto = await _service.CreateAsync(requestDto);
        string createdUrl = Url.Action(nameof(GetDetail), nameof(OrderController), new { id = responseDto.Id }) ?? "";
        return Created(createdUrl, responseDto);
    }

    [HttpPut("{id:int}")]
    [ProducesResponseType<OrderDetailResponseDto>(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
    public async Task<IActionResult> Update([FromRoute] int id, [FromBody] OrderUpsertRequestDto requestDto)
    {
        return Ok(await _service.UpdateAsync(id, requestDto));
    }

    [HttpPut("{id:int}/finish")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
    public async Task<IActionResult> Finish([FromRoute] int id)
    {
        await _service.FinishAsync(id);
        return Ok();
    }

    [HttpDelete("{id:int}")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
    public async Task<IActionResult> Delete([FromRoute] int id)
    {
        await _service.DeleteAsync(id);
        return Ok();
    }

    [HttpGet("default-list-parameters")]
    [ProducesResponseType<OrderListRequestDto>(StatusCodes.Status200OK)]
    public async Task<IActionResult> GetDefaultListParameters()
    {
        await Task.CompletedTask;
        return Ok(new OrderListRequestDto());
    }
    #endregion
}
 